### ISSUE TITLE
Investigate semantic alternatives for generic tags

### ISSUE THREAD

**Issue by tillprochaska** [tactilenews/100eyes#439]
https://github.com/tactilenews/100eyes/pull/432#issuecomment-712334103

**Comment by mattwr18** [comment]
Yes! Thanks for opening this ticket. I think we could extend the scope of this ticket to also include the other concern I brought up, `divitis`, otherwise we could open a second ticket to discuss that? I will wait to respond to your comment https://github.com/tactilenews/100eyes/pull/432#issuecomment-712334103 until we've decided where to further that convesation.

As far as this topic, semantic HTML, I think we can and should continue to educate ourselves on the semantic alternatives to generic tags, as you put it. The documentation for this topic is vast, and has changed considerably in recent years, I guess. As I pointed out in the `PR` review, `divs`, according to the [MDN docs on %div](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/div) 

>  should be used only when no other semantic element (such as %article or %nav) is appropriate.

And this is by far not the only place this type of declaration can be found, for example take the [W3C docs on %div](https://www.w3.org/TR/2011/WD-html5-author-20110809/the-div-element.html), where they point out the accessibility issues, but also mention the maintainability issue as well.

> NOTE: Authors are strongly encouraged to view the div element as an element of last resort, for when no other element is suitable. Use of the div element instead of more appropriate elements leads to poor accessibility for readers and poor maintainability for authors.

A resource I find helpful is [this blog - Stop using so many divs! An intro to semantic HTML ](https://dev.to/kenbellows/stop-using-so-many-divs-an-intro-to-semantic-html-3i9i), which also has a second parter [Why I care about the Semantic Web](https://dev.to/kenbellows/why-i-care-about-the-semantic-web-2kn7)

I think the accessibility concerns is enough of an argument for us to prioritize this! I think it's important to point out that `a11y` concerns are not only about users with disabilities, as explained in the [MDN docs on accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility)

> However, it is wrong to say that accessibility is just about disabilities. Really, the aim of accessibility is to make your websites/apps usable by as many people in as many contexts as possible, not just those users using high-powered desktop computers.

However, the issue is not only related to `a11y` as pointed out by `W3C` :point_up_2: ... It also makes the site easier to maintain for our future selves and other developers. I found this [StackOverflow post](https://stackoverflow.com/questions/1729447/what-are-the-benefits-of-using-semantic-html) interesting. One poster pointed out that we generally try to get in the habit of using descriptive variables and function names instead of generic ones so that it is easier for us to reason about what the variable/function is without inspecting its content to discover it.

I could go on, but I don't feel like there is opposition to this direction, just that maybe we've gotten in the habit of quickly building the site without taking the time to ask ourselves some basic questions about the semantics. I'm also interested in exploring the specifics of the concept you pointed out in your comment about [composition over inheritance](https://every-layout.dev/rudiments/composition/) and the general structure of using components in different pages with different contexts, and how that fits in with using semantic HTML.

I have found that it is not always straightforward to decide when I should favor using one semantic tag over another. I think I am still making mistakes with this, but I think learning about this and getting a good sense of it is worth the effort. One last resource I have found helpful is this infographic, which helps decide what tag should be used. 
[h5d-sectioning-flowchart.pdf](https://github.com/tactilenews/100eyes/files/5432687/h5d-sectioning-flowchart.pdf)

**Comment by tillprochaska** [comment]
Hi @mattwr18, I haven’t had the time to read everything you wrote in depth, but I wanted to get back to you regarding your concerns about screen reader issues.

Here’s a [screen recording](https://www.dropbox.com/s/w3vpa9yb2a78d17/Screen%20Recording%202020-10-26%20at%2013.26.32.mov?dl=0) how VoiceOver reads out the section of the page you mentioned. The additional, non-semantic wrapping elements don’t add any semantic information, but they also don’t break anything.

Actually, that’s exactly what non-semantic elements are for: They don’t represent anything content-wise, but can be used for styling purposes, grouping etc.

As I said, in the metrics example, you could certainly improve the screen reader output by providing a natural-language variant of what’s displayed (e.g. "1 of 2 have replied" instead of "1/2 have replied") which would make the text that is read out a bit more fluent (see [this screen recording](https://www.dropbox.com/s/08flmd26ka3lhb1/Screen%20Recording%202020-10-26%20at%2013.30.23.mov?dl=0)), but that’s not an issue caused by the wrapping containers.

**Comment by tillprochaska** [comment]
> I think the accessibility concerns is enough of an argument for us to prioritize this! I think it's important to point out that a11y concerns are not only about users with disabilities, as explained in the MDN docs on accessibility

Just let me make clear that I completely agree with you here. I think the main aspect where we’re either misunderstanding each other or simply have different opinions is what to do about *additional* elements.

So in case that we currently use a non-semantic element where a semantic alternative is available, let’s use that. I could imagine a few cases where we could directly apply this, e.g. for the `ChatMessagesGroup` component (which could probably use an `article` element instead of a `div`).

For use cases where we use non-semantic elements for presentational purposes, I still think we should try to reduce the number of wrappers if possible. For example, this is something we use for the onboarding pages:

```erb
<div class="text-serif">
  <%= c 'onboarding_email_form' %>
</div>
```

This could be easily refactored (using a helper in our `ApplicationComponent` as discussed in #437) to something like this, which would simply append the utility class `text-serif` to the components classes:

```erb
<%= c 'onboarding_email_form', class: 'text-serif' %>
```

What I really wouldn’t like to do is eliminate `div`s (in cases where they serve a purely presentational purpose) at the cost of maintainability, just for the sake of reducing the number of elements.
